### PR TITLE
[Cmake] implement (some) ninja support

### PR DESCRIPTION
**Pull request by fuzzard**
## Description
Implements the ability to use ninja to build (some) target projects

## Motivation and context
With some recent changes to adaptive build types, @enen92 alerted me to the fact ninja no longer worked. It only actually worked out of sheer luck really, but this should bring it back for (some) platforms.

Cmake internal lib builds have been standardised to use an IMPORTED library target (eg CrossGUID::CrossGUID). This is done as externalproject_add targets cant easily be added to an imported target. We worked around this previously by using a property list to force add the  externalproject target to libkodi to build. This is a more "Cmake" way i believe

## How has this been tested?
All the following was carried out with the following internal build flags
```
-DENABLE_INTERNAL_FSTRCMP=OFF -DENABLE_INTERNAL_TAGLIB=ON -DENABLE_INTERNAL_SPDLOG=ON -DENABLE_INTERNAL_FMT=ON -DENABLE_INTERNAL_CROSSGUID=ON
```
Build xcode ios project
Build macos x86_64 Debug/Release Unix Makefiles
Build macos x86_64 Debug/Release Ninja
Build aarch64 android Debug Ninja (libkodi.so target, apk target doesnt work)

## What is the effect on users?
N/A

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by fuzzard** [comment]
Ok, have reworked the byproduct stuff. relocated the debug postfix handling to the BUILD_DEP_TARGET. Its a little more relevant to that stuff than the VERSION file parsing.

One thing ive found, generator expressions for VS dont work with cmake < 3.20. Ive increased the cmake version requirement for windows from 3.15 to 3.20. Of note, VS 2019 (community edition) ships with cmake 3.20, so theres not really a need to use a standalone cmake install.

The generator expressions are only used for multi-config based projects (ie Xcode, VS). So this doesnt affect linux users at all, so cmake version requirement can stay where it is. i'll mention ninja does have a multiconfig variant, short answer is i havent tested it, assume it doesnt work, dont be stupid and try to use it.

I need to get approval to remove (or update PATH) to get rid of the standalone cmake 3.15.5 install we have on our jenkins agents. Until i get that confirmation/action, the windows agents will all fail.

**Review by wsnipex — APPROVED** [review]

**Comment by fuzzard** [comment]
@wsnipex sorry to give you grief. Ive gone against my one rule (keep the PR's simple and small). 
Ive reworked the FindFmt and FindSpdlog in 2 new commits. Can you have a look again, particularly at the last 2 commits.

Essentially on some systems (@enen92 ) its still picking up system includes for fmt when using enable_internal_spdlog and fmt. 
The changes now force internal fmt on when internal spdlog is done. Its just easier due to fmt backward incompatibilities.
Both internal spdlog and fmt will now do a find_package call looking for an existing install, and test against our version supplied in the VERSION file in tools/depends. if its older, it'll build internally, if not, it will just populate the include/lib variables and not create the build target.

This would eventually have to have been done, but what this allows is a system to build internal libs once (gets installed into the DEPENDS_PATH location), and then on subsequent cmake generation/run, it'll find those prebuilt libs and just skip building, as long as the version is up to date.
With cmake built libs (ie fmt/spdlog) we can use the cmake config files created/installed more easily on all platforms (ie windows) for the find_package without having to deal with pkgconfig. When we get to autoconf based libs this will get more difficult for windows, but baby steps.
Im not sure if we want to do a force rebuild, but really this should be the most common path. build once, reuse unless VERSION file increases. A force rebuild adds extra complexity, when a user can just clean out the DEPENDS_PATH location to get a fresh start. 

This also is essentially how i see the system handling rechi's comment at https://github.com/xbmc/xbmc/pull/21311#issuecomment-1109457660

**Comment by wsnipex** [comment]
to make reusing stuff from DEPENDS_PATH work for non-depends builds, we need to add $DEPENDS_PATH to CMAKE_PREFIX_PATH

**Comment by fuzzard** [comment]
Yeah, I was looking at that last night. Trying to work out how/where it if https://github.com/xbmc/xbmc/blob/139e39995e9ffe38b583aef84dde0b1f03fa3deb/CMakeLists.txt#L15 is ever used. As far as I can tell, it's not. So I'd look at changing that to depends_path essentially. Would probably have to relocate it a little later (probably after pathsetup)

**Comment by wsnipex** [comment]
that's something for addons afaik.
Maybe something like https://github.com/wsnipex/xbmc/blob/57e4a903a7a39321ed14ac691e0fe3b83f4b71bc/cmake/scripts/common/ArchSetup.cmake#L56-L58 is enough. It'd be nicer to only do that if at least 1 ENABLE_INTERNAL_XX is enabled, but I have no idea how to best detect that

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,7 +1,7 @@
 cmake_minimum_required(VERSION 3.12)
 if(WIN32)
-  # Version 3.15 is required to use "PREPEND" for dependencies
-  cmake_minimum_required(VERSION 3.15)
+  # Version 3.20 is required for multi-config generator expressions to work
+  cmake_minimum_required(VERSION 3.20)
```
Comment (enen92) [inline CMakeLists.txt:4]: it will need changes to the build guides and jenkins workers it seems

Comment (fuzzard) [reply CMakeLists.txt:4]: Doc change is done. It's windows only.

**Comment by wsnipex** [comment]
spdlog/fmt seems ok now, but internal ffmpeg is broken. FFmpeg is built, but the final link fails. On 2nd cmake run ffmpeg will be picked up from DEPENDS_PATH as "external" and it builds fine.

**Comment by fuzzard** [comment]
Is this ffmpeg failure when using ninja, or as a makefile based build? Or both?

**Comment by fuzzard** [comment]
OK, the ffmpeg issue should be fixed now. Missed an update to the wrapper script to use DEPENDS_PATH for PKG_CONFIG_PATH.

Hopefully thats the last for this PR. Still trying to get the windows jenkins slaves "fixed", but thats not a PR issue

**Review by wsnipex — APPROVED** [review]

**Review by howie-f — APPROVED** [review]
tested on Debian 11 with cmake 3.18.4
configurations Release and Debug work with both generators, Unix Makefiles and Ninja.
apart from the windows build slaves i consider this ready now.
cheers @fuzzard. i know you now curse day and hour you first touched cmake's files :)

(empty commit https://github.com/xbmc/xbmc/pull/21463/commits/6ff10075a0115268414cadbace3e6ef7ed4872d4 can be dropped)

File: cmake/modules/FindFFMPEG.cmake [hunk 2]
```diff
@@ -109,12 +109,12 @@ macro(buildFFMPEG)
 "#!${BASH_COMMAND}
 if [[ $@ == *${APP_NAME_LC}.bin* || $@ == *${APP_NAME_LC}${APP_BINARY_SUFFIX}* || $@ == *${APP_NAME_LC}.so* || $@ == *${APP_NAME_LC}-test* ]]
 then
-  avformat=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libavcodec`
-  avcodec=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libavformat`
-  avfilter=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libavfilter`
-  avutil=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libavutil`
-  swscale=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libswscale`
-  swresample=`PKG_CONFIG_PATH=${CMAKE_BINARY_DIR}/${CORE_BUILD_DIR}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libswresample`
+  avformat=`PKG_CONFIG_PATH=${DEPENDS_PATH}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static libavcodec`
```
Comment (wsnipex) [inline cmake/modules/FindFFMPEG.cmake:112]: I know it was wrong before, but this should be `avformat`
and `avcodec` below

**Review by enen92 — APPROVED** [review]
Been using this as my base branch for quite some days :) thanks much for the fix

File: cmake/modules/FindFFMPEG.cmake (hunk 2 above)
Comment (fuzzard) [reply cmake/modules/FindFFMPEG.cmake:112]: Are you sure? testing on osx i get this 

```
MacBook-Pro-2 ~ % PKG_CONFIG_PATH=/Users/Shared/xbmc-depends/macosx12.3_x86_64-target-debug/lib/pkgconfig /Users/Shared/xbmc-depends/arm-darwin21.4.0-native/bin/pkg-config --libs --static avformat
Package avformat was not found in the pkg-config search path.
Perhaps you should add the directory containing `avformat.pc'
to the PKG_CONFIG_PATH environment variable
No package 'avformat' found
MacBook-Pro-2 ~ % PKG_CONFIG_PATH=/Users/Shared/xbmc-depends/macosx12.3_x86_64-target-debug/lib/pkgconfig /Users/Shared/xbmc-depends/arm-darwin21.4.0-native/bin/pkg-config --libs --static libavformat
-L/Users/Shared/xbmc-depends/macosx12.3_x86_64-target-debug/lib -lavformat -lm -lbz2 -lz -lgnutls -lgmp -ldl -lhogweed -lgmp -lnettle -lavcodec -lm -llzma -ldav1d -lz -framework AudioToolbox -pthread -framework VideoToolbox -framework CoreFoundation -framework CoreMedia -framework CoreVideo -framework CoreServices -lswresample -lm -lavutil -pthread -lm -framework VideoToolbox -framework CoreFoundation -framework CoreMedia -framework CoreVideo -framework CoreServices
brent@Brents-MacBook-Pro-2 ~ % 
```

Comment (wsnipex) [reply cmake/modules/FindFFMPEG.cmake:112]: sorry, ofc it's libavformat.
point was:
 **avformat**`=`PKG_CONFIG_PATH=${DEPENDS_PATH}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static **libavformat**
vs
avformat=PKG_CONFIG_PATH=${DEPENDS_PATH}/lib/pkgconfig ${PKG_CONFIG_EXECUTABLE} --libs --static _libavcodec_

**Comment by enen92** [comment]
Jenkins build this please

**Comment by fuzzard** [comment]
No point in that, working through jenkins things still.

**Comment by fuzzard** [comment]
Jenkins build this please

**Pull request merged**